src/Roave/SecurityAdvisories/Version.php

```diff
@@ -35,26 +35,31 @@ public static function fromString(string $version) : self
             throw new \InvalidArgumentException(sprintf('Given version "%s" is not a valid version string', $version));
         }
 
-        return new self($version);
+        return new self(self::stripEndZeroes($version));
     }
 
     public function equalTo(self $other) : bool
     {
-        return (bool) version_compare($this->stripEndZeroes(), $other->stripEndZeroes(), '==');
+        return (bool) version_compare((string)$this, (string) $other, '==');
     }
 
     public function isGreaterThan(self $other) : bool
     {
-        return (bool) version_compare($this->version, $other->version, '>');
+        return (bool) version_compare((string)$this, (string) $other, '>');
     }
 
     public function isGreaterOrEqualThan(self $other) : bool
     {
-        return (bool) version_compare($this->version, $other->version, '>=');
+        return (bool) version_compare((string)$this, (string) $other, '>=');
     }
 
-    public function stripEndZeroes() : string
+    public static function stripEndZeroes(string $version) : string
     {
-        return preg_replace('/\.[\.0+]+$/', '', (string)$this->version);
+        return preg_replace('/\.[\.0+]+$/', '', $version);
+    }
+
+    public function __toString()
```

Comment:
Suggestion: keep this an explicit `->toString()` instead of using an implicit `(string)` cast.

The reason is that a `(string)` cast is a too lax operation, which may lead to bugs hiding and showing up much later. Keeping an explicit `->callSomething()` means:

 * it must be an object
 * it must be a type supporting the `->callSomething()` operation

This is much stricter, and reduces the amount of decisions done by type checker and the reader.

Reply:
thanks a lot for explanation, will do as per your suggestion.